lib/propolis/src/block/crucible.rs

```diff
@@ -57,23 +65,28 @@ impl CrucibleBackend {
 
         volume.activate(gen)?;
 
-        let mut be = Self {
-            block_io: volume.clone(),
-            block_size: 0,
-            sectors: 0,
-            read_only,
-            driver: Mutex::new(None),
-        };
-
         // After active negotiation, set sizes
-        be.block_size =
+        let block_size =
             tokio::task::block_in_place(|| volume.get_block_size())?;
-
         let total_size = tokio::task::block_in_place(|| volume.total_size())?;
+        let sectors = total_size / block_size;
+
+        // TODO: make this tunable?
+        let worker_count = NonZeroUsize::new(8).unwrap();
 
-        be.sectors = total_size / be.block_size;
+        Ok(Arc::new_cyclic(|me| Self {
```

Comment:
I want to know more about this

Reply:
Ok, after some reading, is this to avoid the possibility of memory leaks if something fails during updates to the new `CrucibleBacked`?

Reply:
Peanut gallery comment: my read is that this lets us get rid of a heap allocation (the `Box<BackendProcessFn>` in `struct Driver`) by making the processing callback part of the `Backend` trait that this type already implements. The catch is that this creates a reference cycle (`CrucibleBackend -> Driver -> CrucibleBackend`) where there was none before (we had `CrucibleBackend -> BlockIO` and `CrucibleBackend -> Driver -> BackendProcessFn -> BlockIO`), so we need a weak reference to break it.

I think this simplification applies even if something later fails in the `CrucibleBackend` (since this cycle is set up in `CrucibleBackend::create` it applies no matter whether a subsequent call to something like `attach` succeeds or fails).

Reply:
Yeah, I didn't want creation of the driver to be a separate step.